### PR TITLE
Gutenberg v5.9.x and Contact Form Block Support

### PR DESCRIPTION
**Pull request by gwwar**
#### Changes proposed in this Pull Request:

When using Gutenberg v5.9.0, the Jetpack contact form is not available. Existing blocks display the following in the editor:

Screen Shot 2019-06-12 at 3 11 48 PM

This PR wraps the save functions in a wrapper to satisfy this check: https://github.com/WordPress/gutenberg/pull/14529/files#diff-33e0ac9c44b16a617039a563d4c065f3R108

#### Testing instructions:
- Spin up a test instance with https://jurassic.ninja/create/?jetpack-beta&branch=fix/contact-form-g-5-9
- Activate Contact Form in Jetpack Settings:

Screen Shot 2019-06-12 at 4 08 35 PM

- Install Gutenberg v5.9.0 and activate
- See if contact form is available in the inserter
- No regressions in behavior
- No error in console about `The "save" property must be a valid function`

#### Proposed changelog entry for your changes:

* no changelog entry needed

### CODE REVIEW

**Comment by gwwar** [comment]
@youknowriad @gziolo @aduth Is the original an acceptable usage, or do folks always need to wrap in a function?

File: extensions/blocks/contact-form/index.js
Comment (gwwar) [inline extensions/blocks/contact-form/index.js]: Seeing same behavior here with function vs fat arrow, so this gets past the validation check, but there are regressions in behavior.

**Comment by mmtr** [comment]
Verified the issue on my local Jetpack site (`jetpack/contact-form` cannot be inserted after installing Gutenberg v5.9.0) and tested that changes on this PR make the block insertable again.

However, I noticed more regressions possibly caused by the latest version of the Gutenberg plugin. It seems that all the Jetpack blocks using `InnerBlocks.Content` like `jetpack/contact-form` or `jetpack/contact-info` are currently broken, and none of the changes done to the blocks in the editor are saved to the post content.

**Comment by mmtr** [comment]
Interesting. After changing some imports to pull out the dependencies from `@wordpress/block-editor` instead of from the deprecated `@wordpress/editor`, the regression I was experiencing has been fixed.

Maybe changes introduced in https://github.com/WordPress/gutenberg/pull/15989 have caused the issue and we cannot longer import `InnerBlocks` from `@wordpress/editor`?

**Comment by simison** [comment]
Since Jetpack supports current-1 WordPress version, we need to be careful on relying `@wordpress/block-editor` as it might not be available.

@jeherve how does the core version support for the v7.5 look like?

**Comment by jeherve** [comment]
@simison You're correct. Jetpack supports WP current and -1.

> how does the core version support for the v7.5 look like?

We'll need to keep supporting WP 5.1 in Jetpack until WP 5.3 is released, so Jetpack 7.5 will need to support WP 5.1.

Since `@wordpress/block-editor` was added in WP 5.2 ([ref](https://core.trac.wordpress.org/ticket/46429)), we can't rely on it just yet.

If this bug impacts all blocks relying on Inner Blocks, would it be an option to fix the issue in Gutenberg instead, so all plugins currently using Inner Blocks would be fixed at once?

File: extensions/blocks/contact-form/index.js
Comment (youknowriad) [inline extensions/blocks/contact-form/index.js]: It does seem to me that this is a valid usage as we always considered `save` as a "component". Should we update the check function @gziolo?

Comment (gziolo) [reply extensions/blocks/contact-form/index.js]: In `getSaveElement ` we have https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/serializer.js#L68-L70:

```js
	// Component classes are unsupported for save since serialization must
	// occur synchronously. For improved interoperability with higher-order
	// components which often return component class, emulate basic support.
```

We provide a workaround, so we should relax this check. I will open PR soon.

Comment (gziolo) [reply extensions/blocks/contact-form/index.js]: It seems like this can't be the issue as we have the same check for `save` method almost forever:
https://github.com/WordPress/gutenberg/blob/da8184a055d8b2d40ba0306faf7103c0a961726a/blocks/api/registration.js#L96-L101
```js
if ( ! settings || ! isFunction( settings.save ) ) {
```

Comment (youknowriad) [reply extensions/blocks/contact-form/index.js]: @gziolo Oh, that's a bigger issue that we may even consider for `5.2.x`

Comment (gziolo) [reply extensions/blocks/contact-form/index.js]: Opened https://github.com/WordPress/gutenberg/pull/16152 to address the issue where statics were not hoisted for deprecated components. This should fix the issue which happens here.

**Comment by gwwar** [comment]
Nice Sleuthing @mmtr, so there were difference between the current package and the deprecated export.

> Since Jetpack supports current-1 WordPress version, we need to be careful on relying @wordpress/block-editor as it might not be available.

Sounds like we'll want to audit InnerBlocks usage and then use a conditional `require` vs import until  InnerBlocks is fully removed from the editor package.

**Comment by kwight** [comment]
I'm unclear what next steps are here; this PR is still required, correct? Until the Gutenberg fix lands in the next release?

**Comment by simison** [comment]
If this is fixed in Gutenberg, can we rely on deprecated (and fixed) version of `@wordpress/editor` so that this would work in all current and future WordPress core versions?

That's for the plugin release of course — on wpcom side we can rely on `@wordpress/block-editor` without issues.

**Comment by kwight** [comment]
> The following might work

@gwwar that's slick, and easy to clean up down the road <3

**Comment by ivan-ottinger** [comment]
Hey everyone, just had a concierge session with Business user and the Content form block on their Atomic site is broken.

I tried to deactivate the Gutenberg plugin temporarily, but the issue is the user is already using features of the new version - and therefore deactivating the plugin breaks their existing pages. I am guessing downgrade wouldn't help here either.

Do we have any estimates when the issue could be fixed?

**Comment by youknowriad** [comment]
To update here, we're hoping to release a Gutenberg 5.9.1 release with a fix tomorrow at most.

**Comment by gwwar** [comment]
Oh if 5.9.1 is going out tomorrow, we can hold off on the conditional requires. 

We can leave this open, then close tomorrow as a no-op unless we want D29394-code landing on wpcom to unblock version updates.

File: extensions/blocks/contact-form/index.js
Comment (gziolo) [reply extensions/blocks/contact-form/index.js]: In my testing, you will have to use function expression for this block for safety. It's fragile to assume that `InnerBlocks.Content` will always be a function. The same applies to all `save` implementations. Even if we allow objects, it still might work differently than intended.

Related Slack discussion on WordPress.org: https://wordpress.slack.com/archives/C02QB2JS7/p1560500199068200

**Comment by gwwar** [comment]
Closing this since v5.9.2 of Gutenberg was released

File: extensions/blocks/contact-form/index.js
Comment (youknowriad) [reply extensions/blocks/contact-form/index.js]: @gwwar I think we might still need this change in JetPack because the component is not a function anymore.

**Comment by gwwar** [comment]
Reopening since this is still broken with v5.9.2

**Comment by gwwar** [comment]
Did a quick audit, and it looks like this is the only block that doesn't pass an explicit save function.

**Comment by gwwar** [comment]
@Automattic/jetpack-crew This is ready for review. I think we need a point release to fix this for folks using the Gutenberg plugin. I'll be proposing process updates for next time to avoid this in the future.

**Review by jeherve — APPROVED** [review]
This works well for me, and there are no other blocks impacted by this. Merging.

**Comment by jeherve** [comment]
Cherry-picked to `branch-7.4` in a22fe7f64efabe4826bff5f55380951cfa17d09c

**Comment by gwwar** [comment]
Thanks @jeherve !

**Pull request merged**
